File: Platforms/LeetCode/challenges/may2020/SuperPalindrome.java
public class SuperPalindrome {

    static class Solution {

        boolean isPalindrome(long number) {
            long num=number, n=0;
            while (num!=0) {
                n = n*10 + num%10;
                num /= 10;
            }
            return n==number;
        }

        boolean isPalindrome2(long number) {
            long num=number, n=0;
            long sqr, square, s=0;
            square = number * number;
            sqr = square;

            while (true) {
                if (num==0)
                    break;
                
                if (num!=0) {
                    n = n*10 + num%10;
                    num /= 10;
                }
                if (sqr!=0) {
                    s = s*10 + sqr%10;
                    sqr /= 10;
                }
            }
            if (n!=number)
                return false;
            while (sqr!=0) {
                s = s*10 + sqr%10;
                sqr /= 10;
            }
            return s==square;
        }

        public int superpalindromesInRange(String sL, String sR) {
            long L = Long.valueOf(sL);
            long R = Long.valueOf(sR);
            int MAGIC = 100000;
            int ans = 0;
    
            // count odd length;
            for (int k = 1; k < MAGIC; ++k) {
                StringBuilder sb = new StringBuilder(Integer.toString(k));
                for (int i = sb.length() - 2; i >= 0; --i)
                    sb.append(sb.charAt(i));
                long v = Long.valueOf(sb.toString());
                v *= v;
                if (v > R) break;
                if (v >= L && isPalindrome(v)) ans++;
            }
    
            // count even length;
            for (int k = 1; k < MAGIC; ++k) {
                StringBuilder sb = new StringBuilder(Integer.toString(k));
                for (int i = sb.length() - 1; i >= 0; --i)
                    sb.append(sb.charAt(i));
                long v = Long.valueOf(sb.toString());
                v *= v;
                if (v > R) break;
                if (v >= L && isPalindrome(v)) ans++;
            }
    
            return ans;
        }
    }

    public static void main(String[] args) {
        // System.out.println("Palindrome " + new Solution().isPalindrome(121));
        System.out.println("Number of super-palindromes: " + new Solution().superpalindromesInRange("4", "1000"));
    }
    
}

File: Platforms/LeetCode/challenges/may2020/superPalindrome.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int isPalindrome2(long number) {
    long num=number, n=0;
    long sqr, square, s=0;
    square = number * number;
    sqr = square;

    while (1) {
        if (num==0)
            break;
        
        if (num!=0) {
            n = n*10 + num%10;
            num /= 10;
        }
        if (sqr!=0) {
            s = s*10 + sqr%10;
            sqr /= 10;
        }
    }
    if (n!=number)
        return 0;
    while (sqr!=0) {
        s = s*10 + sqr%10;
        sqr /= 10;
    }
    return s==square;
}

int isPalindrome(long number) {
    long num=number, n=0;
    while (num!=0) {
        n = n*10 + num%10;
        num /= 10;
    }
    return n==number;
}
int superpalindromesInRange(char * left, char * right){
    
    int count = 0;
    long sqrt1 = sqrt(atol(left));
    long sqrt2 = sqrt(atol(right));

    for (long i=sqrt1; i<=sqrt2; ++i) {
        if (i>10 && (i%10!=1 && i%10!=2))
            continue;
        // if (isPalindrome(i) && isPalindrome(i*i)) {
        if (isPalindrome2(i)) {
            // printf("\ni, i*i: %ld, %ld", i, i*i);
            count++;
        }
            
    }
    return count;
    return 0;
}

int main() {
    int count = superpalindromesInRange("4", "999999999999999999");
    printf("\nnumber of super-palindromes in range: %d\n", count);
    /*
    "38455498359"
"999999999999999999"*/
    return 0;
}